### PR TITLE
feat(TCOMP-2598): theme icons support

### PR DESCRIPTION
**Pull request by undx**
https://jira.talendforge.org/browse/TCOMP-2598

### CODE REVIEW

File: component-server-parent/component-server/src/main/java/org/talend/sdk/component/server/front/base/internal/RequestKey.java [hunk 1]
```diff
@@ -28,10 +28,13 @@ public class RequestKey {
...
 
-    public RequestKey(final Locale locale, final boolean includeIconContent, final String query) {
+    private final String theme;
+
+    public RequestKey(final Locale locale, final boolean includeIconContent, final String query, final String theme) {
         this.locale = locale;
         this.includeIconContent = includeIconContent;
         this.query = query;
+        this.theme = theme;
         this.cacheHash = Objects.hash(locale, includeIconContent, query);
```
Comment (ozhelezniak-talend) [inline component-server-parent/component-server/src/main/java/org/talend/sdk/component/server/front/base/internal/RequestKey.java:38]: You didn't change hashcode calculation, is it expected to not include it into hash, but check it in "equals"?

File: component-server-parent/component-server/src/main/java/org/talend/sdk/component/server/service/IconResolver.java [hunk 2]
```diff
@@ -104,38 +119,59 @@ public Icon resolve(final Container container, final String icon) {
...
-                .computeIfAbsent(icon,
-                        k -> ofNullable(getOverridenIcon(icon, appLoader)
-                                .orElseGet(() -> doLoad(container.getLoader(), icon)
-                                        .orElseGet(() -> doLoad(appLoader, icon).orElse(null)))))
-                .orElse(null);
+        if (isThemeSupported) {
+            final String appliedTheme = theme != null ? theme : defaultTheme;
+            final String themedIcon = appliedTheme + "/" + icon;
+            Icon themed = cache.icons
+                    .computeIfAbsent(icon + appliedTheme,
```
Comment (ozhelezniak-talend) [inline component-server-parent/component-server/src/main/java/org/talend/sdk/component/server/service/IconResolver.java:126]: Can we use a "themedIcon" as a key in this Map?

File: component-tools/src/main/java/org/talend/sdk/component/tools/ComponentValidator.java [hunk 3]
```diff
@@ -156,23 +160,52 @@ private String validateIcon(final Icon annotation, final Collection<String> erro
...
-            final Set<File> svgs =
-                    of(classes).map(it -> new File(it, "icons/" + icon + ".svg")).filter(File::exists).collect(toSet());
-            if (svgs.isEmpty()) {
-                log.error("No 'icons/" + icon + ".svg' found, this will run in degraded mode in Talend Cloud");
+            Set<File> svgs;
+            Set<File> pngs;
+            // legacy checks
+            if (configuration.isValidateLegacyIcons()) {
+                svgs = of(classes)
+                        .map(it -> new File(it, ICONS + icon + ".svg"))
```
Comment (ozhelezniak-talend) [inline component-tools/src/main/java/org/talend/sdk/component/tools/ComponentValidator.java:168]: ICONS here as a constant

File: component-tools/src/main/java/org/talend/sdk/component/tools/ComponentValidator.java [hunk 4]
```diff
@@ -156,23 +160,52 @@ private String validateIcon(final Icon annotation, final Collection<String> erro
...
+                        .map(it -> new File(it, ICONS + icon + "_icon32.png"))
+                        .collect(Collectors.toSet());
             } else {
-                if (configuration.isValidateSvg()) {
-                    errors.addAll(svgs.stream().flatMap(this::validateSvg).collect(toSet()));
-                }
+                // themed icons check
+                List<String> prefixes = new ArrayList<>();
+                of(classes).forEach(s -> {
+                    prefixes.add(s + "/icons/light/" + icon);
```
Comment (ozhelezniak-talend) [inline component-tools/src/main/java/org/talend/sdk/component/tools/ComponentValidator.java:177]: and here it's hardcoded, but the the level of the constant is the same

File: talend-component-maven-plugin/src/main/java/org/talend/sdk/component/maven/IconReporterMojo.java [hunk 5]
```diff
@@ -98,7 +98,7 @@ public void execute() throws MojoExecutionException, MojoFailureException {
                     if (missingIcon == null) {
                         final ClassLoader fallbackLoader = IconReporterMojo.class.getClassLoader();
                         try (final InputStream stream = fallbackLoader.getResourceAsStream("icon/missing.png")) {
-                            missingIcon = toDataUri(new IconResolver.Icon("image/png", IO.readBytes(stream)));
+                            missingIcon = toDataUri(new IconResolver.Icon("image/png", IO.readBytes(stream), "light"));
```
Comment (ozhelezniak-talend) [inline talend-component-maven-plugin/src/main/java/org/talend/sdk/component/maven/IconReporterMojo.java:101]: and here we are using "light" instead of ""

**Review by ozhelezniak-talend — APPROVED** [review]
I suppose LGTM
But I leave a few comments

File: component-server-parent/component-server/src/main/java/org/talend/sdk/component/server/front/base/internal/RequestKey.java (hunk 1 above)
Comment (undx) [reply component-server-parent/component-server/src/main/java/org/talend/sdk/component/server/front/base/internal/RequestKey.java:38]: Oh right, good catch!

File: component-server-parent/component-server/src/main/java/org/talend/sdk/component/server/front/ComponentResourceImpl.java [hunk 6]
```diff
@@ -172,14 +172,18 @@ public class ComponentResourceImpl implements ComponentResource {
...
+    private String defaultTheme;
+
     private final Map<String, Function<ComponentIndex, Object>> componentEvaluators = new HashMap<>();
 
     @PostConstruct
     private void setupRuntime() {
-        log.info("Initializing " + getClass());
-
+        log.info("[setupRuntime] Initializing " + getClass());
+        defaultTheme = configuration.getIconDefaultTheme();
```
[the remark this answers is not in the document]
Comment (undx) [reply component-server-parent/component-server/src/main/java/org/talend/sdk/component/server/front/ComponentResourceImpl.java:182]: component-runtime default configuration state.

File: component-server-parent/component-server/src/main/java/org/talend/sdk/component/server/front/ComponentResourceImpl.java [hunk 7]
```diff
@@ -172,14 +172,18 @@ public class ComponentResourceImpl implements ComponentResource {
...
+        final String themeMode = configuration.getSupportIconTheme() ? "themed" : "legacy";
+        log.info("[setupRuntime] Icon mode: {}; default theme: {}.", themeMode, defaultTheme);
         // preload some highly used data
-        getIndex("en", false, null);
+        getIndex("en", false, null, defaultTheme);
```
[the remark this answers is not in the document]
Comment (undx) [reply component-server-parent/component-server/src/main/java/org/talend/sdk/component/server/front/ComponentResourceImpl.java:186]: We initiate first cache result with initial cs configuration (w/o icons).

File: component-server-parent/component-server/src/main/java/org/talend/sdk/component/server/front/ComponentResourceImpl.java [hunk 8]
```diff
@@ -304,23 +308,25 @@ public StreamingOutput getDependency(final String id) {
 
     @Override
     @CacheResult
-    public ComponentIndices getIndex(final String language, final boolean includeIconContent, final String query) {
+    public ComponentIndices getIndex(final String language, final boolean includeIconContent, final String query,
+            final String theme) {
         final Locale locale = localeMapper.mapLocale(language);
+        final String themedIcon = theme == null ? defaultTheme : theme;
```
[the remark this answers is not in the document]
Comment (undx) [reply component-server-parent/component-server/src/main/java/org/talend/sdk/component/server/front/ComponentResourceImpl.java:314]: previous phases were deliberated. here we handle next calls, they can be w/ null values.

File: component-server-parent/component-server/src/main/java/org/talend/sdk/component/server/front/ComponentResourceImpl.java [hunk 9]
```diff
@@ -304,23 +308,25 @@ public StreamingOutput getDependency(final String id) {
...
         caches.evictIfNeeded(indicesPerRequest, configuration.getMaxCacheSize() - 1);
-        return indicesPerRequest.computeIfAbsent(new RequestKey(locale, includeIconContent, query), k -> {
+        return indicesPerRequest.computeIfAbsent(new RequestKey(locale, includeIconContent, query, themedIcon), k -> {
             final Predicate<ComponentIndex> filter = queryLanguageCompiler.compile(query, componentEvaluators);
             return new ComponentIndices(Stream
-                    .concat(findDeployedComponents(includeIconContent, locale), virtualComponents
+                    .concat(findDeployedComponents(includeIconContent, locale, theme), virtualComponents
```
[the remark this answers is not in the document]
Comment (undx) [reply component-server-parent/component-server/src/main/java/org/talend/sdk/component/server/front/ComponentResourceImpl.java:319]: Tthat's correct, it's a mistake on my side.

File: component-server-parent/component-server/src/main/java/org/talend/sdk/component/server/service/IconResolver.java (hunk 2 above)
Comment (undx) [reply component-server-parent/component-server/src/main/java/org/talend/sdk/component/server/service/IconResolver.java:126]: yep.

File: component-server-parent/component-server/src/main/java/org/talend/sdk/component/server/service/IconResolver.java [hunk 10]
```diff
@@ -104,38 +119,59 @@ public Icon resolve(final Container container, final String icon) {
...
+                                k -> ofNullable(doLoad(container.getLoader(), icon, "")
+                                        .orElseGet(() -> doLoad(appLoader, icon, "")
+                                                .orElse(null))))
+                        .orElse(null);
+            }
+            return themed;
+        } else {
+            return cache.icons
+                    .computeIfAbsent(icon,
+                            k -> ofNullable(getOverridenIcon(icon, appLoader)
```
[the remark this answers is not in the document]
Comment (undx) [reply component-server-parent/component-server/src/main/java/org/talend/sdk/component/server/service/IconResolver.java:144]: Configuration can be:
- theme support + legacy support
- theme support + no legacy support
- no theme support

File: component-tools/src/main/java/org/talend/sdk/component/tools/ComponentValidator.java (hunk 3 above)
Comment (undx) [reply component-tools/src/main/java/org/talend/sdk/component/tools/ComponentValidator.java:168]: Don't get the comment?
`ICONS="icons/"` is the standard prefix for legacy icons.

File: talend-component-maven-plugin/src/main/java/org/talend/sdk/component/maven/IconReporterMojo.java (hunk 5 above)
Comment (undx) [reply talend-component-maven-plugin/src/main/java/org/talend/sdk/component/maven/IconReporterMojo.java:101]: The icon reporter mojo is not more in use now. Just the default theme fallback.

File: component-tools/src/main/java/org/talend/sdk/component/tools/ComponentValidator.java (hunk 4 above)
Comment (undx) [reply component-tools/src/main/java/org/talend/sdk/component/tools/ComponentValidator.java:177]: hum, yay can  introduce the constant.

File: component-server-parent/component-server/src/test/java/org/talend/sdk/component/server/front/ComponentResourceImplTest.java [hunk 11]
```diff
@@ -514,6 +607,7 @@ private void assertComponent(final String plugin, final String family, final Str
         } else if ("chain".equals(data.getId().getFamily())
                 && ("file".equals(data.getId().getName()) || "standalone".equals(data.getId().getName()))) {
             assertEquals("myicon", data.getIcon().getIcon());
+            assertEquals("light", data.getIcon().getTheme());
```
[the remark this answers is not in the document]
Comment (undx) [reply component-server-parent/component-server/src/test/java/org/talend/sdk/component/server/front/ComponentResourceImplTest.java:610]: Yes, a few lines upper in jdbc family;

File: component-tools/src/main/java/org/talend/sdk/component/tools/ComponentValidator.java [hunk 12]
```diff
@@ -156,23 +160,52 @@ private String validateIcon(final Icon annotation, final Collection<String> erro
...
+                        .map(it -> new File(it, ICONS + icon + "_icon32.png"))
+                        .collect(Collectors.toSet());
             } else {
-                if (configuration.isValidateSvg()) {
-                    errors.addAll(svgs.stream().flatMap(this::validateSvg).collect(toSet()));
-                }
+                // themed icons check
+                List<String> prefixes = new ArrayList<>();
+                of(classes).forEach(s -> {
+                    prefixes.add(s + "/" + ICONS + "light/" + icon);
```
Comment (ozhelezniak-talend) [inline component-tools/src/main/java/org/talend/sdk/component/tools/ComponentValidator.java:177]: Why does here it start as an absolute path?
group comment $2.1

File: component-tools/src/main/java/org/talend/sdk/component/tools/ComponentValidator.java [hunk 13]
```diff
@@ -156,23 +160,52 @@ private String validateIcon(final Icon annotation, final Collection<String> erro
...
-                log.error("No 'icons/" + icon + ".svg' found, this will run in degraded mode in Talend Cloud");
+            Set<File> svgs;
+            Set<File> pngs;
+            // legacy checks
+            if (configuration.isValidateLegacyIcons()) {
+                svgs = of(classes)
+                        .map(it -> new File(it, ICONS + icon + ".svg"))
+                        .collect(toSet());
+                pngs = Stream.of(classes)
+                        .map(it -> new File(it, ICONS + icon + "_icon32.png"))
```
Comment (ozhelezniak-talend) [inline component-tools/src/main/java/org/talend/sdk/component/tools/ComponentValidator.java:171]: And here the path is relative (don't start from "/"). When we do the same for themed icons, the path starts from "/"
group comment $2.2

**Review by ozhelezniak-talend — COMMENTED** [review]
I have a question about the difference in pathes for legacy and themed icons.

File: component-tools/src/main/java/org/talend/sdk/component/tools/ComponentValidator.java (hunk 12 above)
Comment (undx) [reply component-tools/src/main/java/org/talend/sdk/component/tools/ComponentValidator.java:177]: Here we're building a complete path string.

File: component-tools/src/main/java/org/talend/sdk/component/tools/ComponentValidator.java (hunk 13 above)
Comment (undx) [reply component-tools/src/main/java/org/talend/sdk/component/tools/ComponentValidator.java:171]: but here we're using the File class wih parent and child parameters, not need to be absolute.

Comment (ozhelezniak-talend) [reply component-tools/src/main/java/org/talend/sdk/component/tools/ComponentValidator.java:171]: Oh, I see now. Thank you

**Review by ozhelezniak-talend — APPROVED** [review]
lgtm

**Pull request merged**